ci: remove deprecated actions and use stable

joshka opened the pull request:
nightly is only necessary for formatting. For most things in git-cliff
stable is sufficient. So use stable for everything else.

actions-rs/toolchain was archived years ago. Generally use
dtolnay/rust-toolchain as a replacement

actions-rs/cargo is similarly archived. Use plain cargo commands

taike-e/install-action automatically uses cargo-binstall to install
tools not setup for install-action directly. So there's no need to
install binstall to install cargo-msrv

## Description

<!--- Describe your changes in detail -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots / Logs (if applicable)

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other <!--- (provide information) -->

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] My code follows the code style of this project.
- [ ] I have updated the documentation accordingly.
- [ ] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [ ] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--- Thank you for contributing to git-cliff! ⛰️  -->

joshka committed:
ci: remove deprecated actions and use stable

nightly is only necessary for formatting. For most things in git-cliff
stable is sufficient. So use stable for everything else.

actions-rs/toolchain was archived years ago. Generally use
dtolnay/rust-toolchain as a replacement

actions-rs/cargo is similarly archived. Use plain cargo commands

taike-e/install-action automatically uses cargo-binstall to install
tools not setup for install-action directly. So there's no need to
install binstall to install cargo-msrv

orhun reviewed (commented):
Thanks, looking great!

I'm not so sure about the changes in `cd.yml` though.

joshka commented on .github/workflows/cd.yml:
```diff
@@ -34,95 +34,83 @@ jobs:
           - {
               NAME: linux-x64-glibc,
               OS: ubuntu-22.04,
-              TOOLCHAIN: stable,
```
With the change to use dtolnay/rust-toolchain, this variable is not used to do anything, so it makes the workflow confusing to have it specified. It's easy to add in the future if needed, but generally YAGNI.

If there is a per combination change that's used once though, a future change could instead do something like `if: ${{ matrix.build.TOOLCHAIN != '' }}` ... instead of repeating the toolchain for each item.

orhun replied:
Yeah, sometimes some deps magically do not work with stable Rust in the CI (it happened before).

It happens during the release so I'd prefer if we can do that now instead of hastily doing that after realizing the release is broken or something.

Sorry if I'm a bit pushy, I apparently have PTSD from one of the past releases lol

joshka commented on .github/workflows/cd.yml:
```diff
@@ -142,11 +130,9 @@ jobs:
-          toolchain: ${{ matrix.build.TOOLCHAIN }}
-          target: ${{ matrix.build.TARGET }}
-          override: true
+          targets: ${{ matrix.build.TARGET }}
```
yes, it overrides future commands that use cargo. I think it may not override the actions/cargo below (which I missed from this PR). From memory that installs rust instead of using the default rust.

joshka committed:
ci: simplify cd.yml

- Use matrix include instead of an object per build
- simplify some conditional logic
- use latest os versions
- use arm versions instead of cross compiling

joshka commented on .github/workflows/cd.yml:
```diff
@@ -26,140 +26,93 @@ jobs:
-              NPM_PUBLISH: true,
-              PYPI_PUBLISH: true,
-            }
+        include:
```
See https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategymatrixinclude

joshka commented on .github/workflows/cd.yml:
```diff
@@ -168,122 +121,119 @@ jobs:
           OUT_DIR=release/man/ cargo run --release --bin git-cliff-mangen
           for bin in 'git-cliff' 'git-cliff-completions' 'git-cliff-mangen'; do
-            if [ "${{ matrix.build.OS }}" = "windows-2022" ]; then
+            if [ "${{ contains(matrix.os, 'windows') }}" = "true" ]; then
```
This can be further simplified and deduplicated by splitting the os and runner in the matrix above, either by adding a `runner` variable, or by adding an `is_windows` variable. I didn't worry about this as it's clear enough here.

There's possibly also a bash test that checks for `true` instead of string comparisonI but I haven't looked.

joshka commented on .github/workflows/cd.yml:
```diff
@@ -26,140 +26,93 @@ jobs:
-          toolchain: ${{ matrix.build.TOOLCHAIN }}
-          target: ${{ matrix.build.TARGET }}
-          override: true
-      - name: Build (linux/macos)
-        if: matrix.build.OS != 'windows-2022'
-        uses: actions-rs/cargo@v1
-        with:
-          use-cross: true
-          command: build
-          args: --release --locked --target ${{ matrix.build.TARGET }}
```
I think cross is no longer necessary because the arm stuff is being built on the right runner now.

I'd suggest doing a quick check to validate this by turning off all the release parts of this change in another PR and just checking that all the build succeeds.

orhun replied:
I'm not so sure about not needing `cross`. This was extremely fragile to put together and testing it takes a lot of time.

joshka replied:
Checking this in https://github.com/orhun/git-cliff/pull/1198

joshka commented on .github/workflows/cd.yml:
```diff
@@ -26,140 +26,93 @@ jobs:
+            publish_npm: false
+            publish_pypi: true
+            install_musl: true
+          - name: linux-arm64-glibc
```
See https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#standard-github-hosted-runners-for-public-repositories for available runners

joshka commented on .github/workflows/cd.yml:
```diff
@@ -168,122 +121,119 @@ jobs:
           body: "${{ needs.generate-changelog.outputs.release_body }}"
       - name: Publish to GitHub (pre-release)
-        if: ${{ contains(github.ref, '-') }}
+        if: contains(github.ref, '-')
```
https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstepsif
>When you use expressions in an if conditional, you can, optionally, omit the ${{ }} expression syntax because GitHub Actions automatically evaluates the if conditional as an expression. However, this exception does not apply everywhere.

orhun replied:
that's a cool shorthand
